configuration.md
---
# !!! DO NOT MODIFY !!!
# This file is auto-generated by the gendoc tool based on the source code of the app.
description: This section describes the configuration parameters and their types for the evil tools.
keywords:
- Evil Tools
- Evil Wallet
- Evil Spammer
- Configuration
- JSON
- Customize
- Config
- reference
---


# Configuration

evil tools use a JSON standard format as a config file. If you are unsure about JSON syntax, you can find more information in the [official JSON specs](https://www.json.org).

You can change the path of the config file by using the `-c` or `--config` argument while executing `evil-tools` executable.

For example:
```bash
evil-tools -c config_example.json
```

You can always get the most up-to-date description of the config parameters by running:

```bash
evil-tools -h --full
```

## <a id="app"></a> 1. Application

| Name                      | Description                                            | Type    | Default value |
| ------------------------- | ------------------------------------------------------ | ------- | ------------- |
| checkForUpdates           | Whether to check for updates of the application or not | boolean | true          |
| [shutdown](#app_shutdown) | Configuration for shutdown                             | object  |               |

### <a id="app_shutdown"></a> Shutdown

| Name                     | Description                                                                                            | Type   | Default value |
| ------------------------ | ------------------------------------------------------------------------------------------------------ | ------ | ------------- |
| stopGracePeriod          | The maximum time to wait for background processes to finish during shutdown before terminating the app | string | "3s"          |
| [log](#app_shutdown_log) | Configuration for log                                                                                  | object |               |

### <a id="app_shutdown_log"></a> Log

| Name     | Description                                         | Type    | Default value  |
| -------- | --------------------------------------------------- | ------- | -------------- |
| enabled  | Whether to store self-shutdown events to a log file | boolean | true           |
| filePath | The file path to the self-shutdown log              | string  | "shutdown.log" |

Example:

```json
  {
    "app": {
      "checkForUpdates": true,
      "shutdown": {
        "stopGracePeriod": "3s",
        "log": {
          "enabled": true,
          "filePath": "shutdown.log"
        }
      }
    }
  }
```

## <a id="logger"></a> 2. Logger

| Name        | Description                                                                                                    | Type   | Default value |
| ----------- | -------------------------------------------------------------------------------------------------------------- | ------ | ------------- |
| name        | The optional name of the logger instance. All log messages are prefixed with that name.                        | string | ""            |
| level       | The minimum enabled logging level                                                                              | string | "info"        |
| timeFormat  | Sets the logger's timestamp format. (options: "rfc3339", "rfc3339nano", "datetime", "timeonly", and "iso8601") | string | "rfc3339"     |
| outputPaths | A list of file paths or stdout/stderr to write logging output to                                               | array  | stdout        |

Example:

```json
  {
    "logger": {
      "name": "",
      "level": "info",
      "timeFormat": "rfc3339",
      "outputPaths": [
        "stdout"
      ]
    }
  }
```

## <a id="tool"></a> 3. Tool

| Name                                | Description                                                                                                                                                                           | Type   | Default value                                                                                                                      |
| ----------------------------------- | ------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- | ------ | ---------------------------------------------------------------------------------------------------------------------------------- |
| nodeURLs                            | API URLs for clients used in test separated with commas                                                                                                                               | array  | http://localhost:8050                                                                                                              |
| faucetURL                           | Faucet URL used in test                                                                                                                                                               | string | "http://localhost:8088"                                                                                                            |
| faucetSplitNumber                   | Number of outputs to split faucet outputs into, this determine number of txs that are sent per one faucet request. Reduce this number if not all txs are accepted due to congestion.  | int    | 80                                                                                                                                 |
| accountStatesFile                   | File to store account states in                                                                                                                                                       | string | "wallet.dat"                                                                                                                       |
| blockIssuerPrivateKey               | Block issuer private key (in hex) to use for genesis account spams                                                                                                                    | string | "db39d2fde6301d313b108dc9db1ee724d0f405f6fde966bd776365bc5f4a5fb31e4b21eb51dcddf65c20db1065e1f1514658b23a3ddbf48d30c0efc926a9a648" |
| accountID                           | Account ID to use for genesis account                                                                                                                                                 | string | "0x6aee704f25558e8aa7630fed0121da53074188abc423b3c5810f80be4936eb6e"                                                               |
| faucetRequestsBlockIssuerPrivateKey | Block issuer private key (in hex) to use for funds preparation from faucet outputs                                                                                                    | string | ""                                                                                                                                 |
| faucetRequestsAccountID             | Account ID to use for fund preparation.                                                                                                                                               | string | ""                                                                                                                                 |

Example:

```json
  {
    "tool": {
      "nodeURLs": [
        "http://localhost:8050"
      ],
      "faucetURL": "http://localhost:8088",
      "faucetSplitNumber": 80,
      "accountStatesFile": "wallet.dat",
      "blockIssuerPrivateKey": "db39d2fde6301d313b108dc9db1ee724d0f405f6fde966bd776365bc5f4a5fb31e4b21eb51dcddf65c20db1065e1f1514658b23a3ddbf48d30c0efc926a9a648",
      "accountID": "0x6aee704f25558e8aa7630fed0121da53074188abc423b3c5810f80be4936eb6e",
      "faucetRequestsBlockIssuerPrivateKey": "",
      "faucetRequestsAccountID": ""
    }
  }
```

## <a id="profiling"></a> 4. Profiling

| Name        | Description                                       | Type    | Default value    |
| ----------- | ------------------------------------------------- | ------- | ---------------- |
| enabled     | Whether the profiling component is enabled        | boolean | false            |
| bindAddress | The bind address on which the profiler listens on | string  | "localhost:6060" |

Example:

```json
  {
    "profiling": {
      "enabled": false,
      "bindAddress": "localhost:6060"
    }
  }
```

## <a id="info"></a> 5. Info

| Name  | Description                                      | Type   | Default value |
| ----- | ------------------------------------------------ | ------ | ------------- |
| alias | Alias for which info command should be executed. | string | ""            |

Example:

```json
  {
    "info": {
      "alias": ""
    }
  }
```

## <a id="accounts"></a> 6. Accounts

| Name                           | Description                | Type   | Default value |
| ------------------------------ | -------------------------- | ------ | ------------- |
| [create](#accounts_create)     | Configuration for create   | object |               |
| [convert](#accounts_convert)   | Configuration for convert  | object |               |
| [destroy](#accounts_destroy)   | Configuration for destroy  | object |               |
| [allot](#accounts_allot)       | Configuration for allot    | object |               |
| [stake](#accounts_stake)       | Configuration for stake    | object |               |
| [claim](#accounts_claim)       | Configuration for claim    | object |               |
| [delegate](#accounts_delegate) | Configuration for delegate | object |               |
| [update](#accounts_update)     | Configuration for update   | object |               |

### <a id="accounts_create"></a> Create

| Name                 | Description                                                                                                                                         | Type    | Default value |
| -------------------- | --------------------------------------------------------------------------------------------------------------------------------------------------- | ------- | ------------- |
| alias                | Alias name of the account to create                                                                                                                 | string  | ""            |
| noBlockIssuerFeature | Create account without Block Issuer Feature, can only be set false no if implicit is false, as each account created implicitly needs to have BIF.   | boolean | false         |
| implicit             | Create an implicit account                                                                                                                          | boolean | false         |
| transition           | Account should be transitioned to a full account if created with implicit address. Transition disabled by default, to enable provide an empty flag. | boolean | false         |

### <a id="accounts_convert"></a> Convert

| Name  | Description                                                   | Type   | Default value |
| ----- | ------------------------------------------------------------- | ------ | ------------- |
| alias | The implicit account to be converted to full account with BIF | string | ""            |

### <a id="accounts_destroy"></a> Destroy

| Name       | Description                                    | Type   | Default value |
| ---------- | ---------------------------------------------- | ------ | ------------- |
| alias      | The alias name of the account to be destroyed  | string | ""            |
| expirySlot | The expiry slot of the account to be destroyed | int    | 0             |

### <a id="accounts_allot"></a> Allot

| Name   | Description                                    | Type   | Default value |
| ------ | ---------------------------------------------- | ------ | ------------- |
| alias  | The alias name of the account to allot mana to | string | ""            |
| amount | The amount of mana to allot                    | uint   | 10000         |

### <a id="accounts_stake"></a> Stake

| Name       | Description                             | Type   | Default value |
| ---------- | --------------------------------------- | ------ | ------------- |
| alias      | The alias name of the account to stake  | string | ""            |
| amount     | The amount of tokens to stake           | int    | 100           |
| fixedCost  | The fixed cost of the account to stake  | int    | 0             |
| startEpoch | The start epoch of the account to stake | int    | 0             |
| endEpoch   | The end epoch of the account to stake   | int    | 0             |

### <a id="accounts_claim"></a> Claim

| Name  | Description                                     | Type   | Default value |
| ----- | ----------------------------------------------- | ------ | ------------- |
| alias | The alias name of the wallet to get rewards for | string | ""            |

### <a id="accounts_delegate"></a> Delegate

| Name      | Description                                                                      | Type    | Default value                                                     |
| --------- | -------------------------------------------------------------------------------- | ------- | ----------------------------------------------------------------- |
| fromAlias | The alias of the delegator, does not need to have an account                     | string  | "default"                                                         |
| toAddress | The account address of the account to delegate IOTA tokens to                    | string  | "rms1pzg8cqhfxqhq7pt37y8cs4v5u4kcc48lquy2k73ehsdhf5ukhya3y5rx2w6" |
| amount    | The amount of mana to delegate                                                   | int     | 100                                                               |
| checkPool | Check if the delegation is added to pool stake when the start epoch is committed | boolean | false                                                             |

### <a id="accounts_update"></a> Update

| Name           | Description                           | Type   | Default value |
| -------------- | ------------------------------------- | ------ | ------------- |
| alias          | Alias name of the account to update   | string | ""            |
| blockIssuerKey | Block issuer key (in hex) to add      | string | ""            |
| amount         | Amount of token to add                | int    | 100           |
| mana           | Amount of mana to add                 | int    | 100           |
| expirySlot     | Update the expiry slot of the account | int    | 0             |

Example:

```json
  {
    "accounts": {
      "create": {
        "alias": "",
        "noBlockIssuerFeature": false,
        "implicit": false,
        "transition": false
      },
      "convert": {
        "alias": ""
      },
      "destroy": {
        "alias": "",
        "expirySlot": 0
      },
      "allot": {
        "alias": "",
        "amount": 10000
      },
      "stake": {
        "alias": "",
        "amount": 100,
        "fixedCost": 0,
        "startEpoch": 0,
        "endEpoch": 0
      },
      "claim": {
        "alias": ""
      },
      "delegate": {
        "fromAlias": "default",
        "toAddress": "rms1pzg8cqhfxqhq7pt37y8cs4v5u4kcc48lquy2k73ehsdhf5ukhya3y5rx2w6",
        "amount": 100,
        "checkPool": false
      },
      "update": {
        "alias": "",
        "blockIssuerKey": "",
        "amount": 100,
        "mana": 100,
        "expirySlot": 0
      }
    }
  }
```

## <a id="spammer"></a> 7. Spammer

| Name                  | Description                                                                                                                                                                                                                                                                        | Type    | Default value |
| --------------------- | ---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- | ------- | ------------- |
| type                  | Spammers used during test. Format: strings separated with comma, available options: 'blk' - block, 'tx' - transaction, 'ds' - double spends spammers, 'nds' - n-spends spammer, 'bb' - blowball, or one of custom scenarios that can be found in pkg/evilwallet/customscenarion.go | string  | "tx"          |
| rate                  | Spamming rate for provided 'spammer'. Format: numbers separated with comma, e.g. 10,100,1 if three spammers were provided for 'spammer' parameter.                                                                                                                                 | int     | 1             |
| duration              | Spam duration. If not provided spam will lats infinitely. Format: separated by commas list of decimal numbers, each with optional fraction and a unit suffix, such as '300ms', '-1.5h' or '2h45m'.
 Valid time units are 'ns', 'us', 'ms', 's', 'm', 'h'.                          | string  | "-1ns"        |
| account               | Account alias to be used for the spam. Account should be created first with accounts tool.                                                                                                                                                                                         | string  | ""            |
| rateSetterEnabled     | Enable the rate setter, which will set the rate for the spammer. To enable provide an empty flag.                                                                                                                                                                                  | boolean | false         |
| deepSpamEnabled       | Enable the deep spam, by reusing outputs created during the spam. To enable provide an empty flag.                                                                                                                                                                                 | boolean | false         |
| reuseEnabled          | Enable the reuse of outputs created during the spam. To enable provide an empty flag.                                                                                                                                                                                              | boolean | false         |
| autoRequestingEnabled | Enable the auto-requesting, which will request tokens from faucet for the spammer. To enable provide an empty flag.                                                                                                                                                                | boolean | false         |
| autoRequestingAmount  | Amount of tokens to be requested from faucet for the spammer. To enable provide an empty flag.                                                                                                                                                                                     | int     | 1000          |
| nSpend                | Number of outputs to be spent in n-spends spammer for the spammer type needs to be set to 'ds'. Default value is 2 for double-spend.                                                                                                                                               | int     | 2             |
| blowballSize          | Size of the blowball to be used in blowball spammer. To enable provide an empty flag.                                                                                                                                                                                              | int     | 30            |

Example:

```json
  {
    "spammer": {
      "type": "tx",
      "rate": 1,
      "duration": "-1ns",
      "account": "",
      "rateSetterEnabled": false,
      "deepSpamEnabled": false,
      "reuseEnabled": false,
      "autoRequestingEnabled": false,
      "autoRequestingAmount": 1000,
      "nSpend": 2,
      "blowballSize": 30
    }
  }
```



tools/gendoc/configuration_header.md
---
description: This section describes the configuration parameters and their types for the evil tools.
keywords:
- Evil Tools
- Evil Wallet
- Evil Spammer
- Configuration
- JSON
- Customize
- Config
- reference
---


# Configuration

evil tools use a JSON standard format as a config file. If you are unsure about JSON syntax, you can find more information in the [official JSON specs](https://www.json.org).

You can change the path of the config file by using the `-c` or `--config` argument while executing `evil-tools` executable.

For example:
```bash
evil-tools -c config_example.json
```

You can always get the most up-to-date description of the config parameters by running:

```bash
evil-tools -h --full
```



config.go
package main

import (
	"time"

	"github.com/iotaledger/evil-tools/pkg/accountwallet"
	"github.com/iotaledger/evil-tools/pkg/evilwallet"
	"github.com/iotaledger/evil-tools/pkg/spammer"
	"github.com/iotaledger/evil-tools/programs"
)

// Nodes used during the test, use at least two nodes to be able to double spend.
var (
	// urls = []string{"http://bootstrap-01.feature.shimmer.iota.cafe:8080", "http://vanilla-01.feature.shimmer.iota.cafe:8080", "http://drng-01.feature.shimmer.iota.cafe:8080"}
	urls = []string{"http://localhost:8050"} //, "http://localhost:8070", "http://localhost:8040"}
)

var (
	customSpamParams = programs.CustomSpamParams{
		ClientURLs:            urls,
		FaucetURL:             "http://localhost:8088",
		SpamType:              spammer.TypeBlock,
		Rate:                  1,
		TimeUnit:              time.Second,
		DelayBetweenConflicts: 0,
		NSpend:                2,
		Scenario:              evilwallet.Scenario1(),
		ScenarioName:          "guava",
		DeepSpam:              false,
		EnableRateSetter:      false,
		AccountAlias:          accountwallet.GenesisAccountAlias,
		BlowballSize:          30,
	}

	accountsSubcommandsFlags []accountwallet.AccountSubcommands
)
